cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizerServiceTest.java

```diff
@@ -0,0 +1,193 @@
+    return cConf;
+  }
+
+  private SConfiguration createSConf() {
+    SConfiguration sConf = SConfiguration.create();
+    return sConf;
+  }
+
+  private ArtifactLocalizerService setupFileLocalizerService(int port) throws IOException {
+    CConfiguration cConf = createCConf(port);
+    SConfiguration sConf = createSConf();
+    DiscoveryServiceClient discoveryClient = getInjector().getInstance(DiscoveryServiceClient.class);
+    RemotePluginFinder remotePluginFinder = getInjector().getInstance(RemotePluginFinder.class);
+    ArtifactLocalizerService artifactLocalizerService =
+      new ArtifactLocalizerService(cConf, sConf, new InMemoryDiscoveryService(),
+                                   new ArtifactLocalizer(discoveryClient,
+                                                         tmpFolder.newFolder()));
+    // start the service
+    artifactLocalizerService.startAndWait();
+
+    return artifactLocalizerService;
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    this.localizerService = setupFileLocalizerService(10001);
+    getInjector().getInstance(ArtifactRepository.class).clear(NamespaceId.DEFAULT);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    this.localizerService.shutDown();
+  }
+
+  @Test
+  public void testUnpackArtifact() throws Exception {
+
+    LocationFactory locationFactory = getInjector().getInstance(LocationFactory.class);
+    ArtifactRepository artifactRepository = getInjector().getInstance(ArtifactRepository.class);
+    ArtifactLocalizerClient client = getInjector().getInstance(ArtifactLocalizerClient.class);
+
+    Id.Artifact artifactId = Id.Artifact.from(Id.Namespace.DEFAULT, "some-task", "1.0.0-SNAPSHOT");
+    Location appJar = AppJarHelper.createDeploymentJar(locationFactory, TaskWorkerServiceTest.TestRunnableClass.class);
+    File appJarFile = new File(tmpFolder.newFolder(),
+                               String.format("%s-%s.jar", artifactId.getName(), artifactId.getVersion().getVersion()));
+    File newAppJarFile = new File(tmpFolder.newFolder(),
+                                  String.format("%s-%s-copy.jar", artifactId.getName(),
+                                                artifactId.getVersion().getVersion()));
+    Files.copy(Locations.newInputSupplier(appJar), appJarFile);
+    appJar.delete();
+    artifactRepository.addArtifact(artifactId, appJarFile);
+
+    Location unpackedDir = client.getUnpackedArtifactLocation(artifactId.toEntityId());
+
+    // Make sure the artifact was actually cached
+    validateUnpackDir(unpackedDir);
+
+    // Call the sidecar again and make sure the same path was returned
+    Location sameUnpackedDir = client.getUnpackedArtifactLocation(artifactId.toEntityId());
+    Assert.assertEquals(unpackedDir, sameUnpackedDir);
+
+    // Delete and recreate the artifact to update the last modified date
+    artifactRepository.deleteArtifact(artifactId);
+
+    Thread.sleep(1000);
```

Comment:
Don't do sleep in unit-test. Why need to sleep?

Reply:
This was just for testing, Its removed now

Reply:
Is it removed??